src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -149,87 +169,158 @@ class FastSync(
         handleRequestFailure(assignedHandlers(ref), ref, "Unexpected error")
     }
 
-    private def removeRequestHandler(handler: ActorRef): Unit = {
-      context unwatch handler
-      assignedHandlers -= handler
+    def waitingForTargetBlockUpdate(state: FinalBlockProcessingResult): Receive = handleCommonMessages orElse {
+      case FastSyncTargetBlockSelector.Result(targetBlockHeader) =>
+        log.info(s"new target block with number ${targetBlockHeader.number} received")
+        if (targetBlockHeader.number >= syncState.targetBlock.number) {
+          updateTargetSyncState(state, targetBlockHeader)
+          syncState = syncState.copy(updatingTargetBlock = false)
+          context become this.receive
+          processSyncing()
+        } else {
+          syncState = syncState.copy(targetBlockUpdateFailures = syncState.targetBlockUpdateFailures + 1)
+          scheduler.scheduleOnce(syncRetryInterval, self, WaitForDownloadsToFinish(state))
+        }
+
+      case PersistSyncState => persistSyncState()
+
+      case WaitForDownloadsToFinish(sState) => updateTargetBlock(sState)
     }
 
-    private def insertHeader(header: BlockHeader): Boolean = {
-      val parentTdOpt = blockchain.getTotalDifficultyByHash(header.parentHash)
-      parentTdOpt foreach { parentTotalDifficulty =>
-        blockchain.save(header)
-        blockchain.save(header.hash, parentTotalDifficulty + header.difficulty)
+    private def updateTargetBlock(state: FinalBlockProcessingResult): Unit = {
+      syncState = syncState.copy(updatingTargetBlock = true)
+      if (syncState.targetBlockUpdateFailures <= syncConfig.maximumTargetUpdateFailures) {
+        if (assignedHandlers.nonEmpty) {
+          log.info(s"Still waiting for some responses, rescheduling target block update")
+          scheduler.scheduleOnce(syncRetryInterval, self, WaitForDownloadsToFinish(state))
+        } else {
+          log.info("Asking for new target block")
+          val targetBlockSelector =
+            context.actorOf(FastSyncTargetBlockSelector.props(etcPeerManager, peerEventBus, syncConfig, scheduler))
+          targetBlockSelector ! FastSyncTargetBlockSelector.ChooseTargetBlock
+          context become waitingForTargetBlockUpdate(state)
+        }
+      } else {
+        log.warning(s"Sync failure! Number of targetBlock Failures reached maximum.")
+        sys.exit(1)
```

Comment:
could we start downloading the state for the best block we currently have, instead of exiting?

Reply:
hmm, we definitely could but I am not sure it would help here, as  we only validate every n-th block, and then every block in interval  (pivot -  X, pivot + X), so to be sure that current best block is safe we would need to discard the last X blocks and then download and validate all blocks to the point from `currentBest - X`  to `currentBest + X.`
This procedure can also fail, so we would be back to declaring a failure at some point in time.